fastfield_codecs/src/null_index/sparse.rs

```diff
@@ -0,0 +1,534 @@
+/// A block consists of max u16 values
+#[derive(Debug)]
+struct SparseCodecBlock {
+    /// The number of values in the block
+    num_vals: u16,
+    /// The number of values set before the block
+    offset: u32,
+}
+
+impl SparseCodecBlock {
+    fn empty_block(offset: u32) -> Self {
+        Self {
+            num_vals: 0,
+            offset,
+        }
+    }
+
+    fn new(num_vals: u16, offset: u32) -> Self {
+        Self { num_vals, offset }
+    }
+
+    #[inline]
+    fn value_at_idx(&self, data: &[u8], idx: u16) -> u16 {
+        let start_offset: usize = (self.offset + idx as u32) as usize * 2;
+        let bytes: [u8; 2] = data[start_offset..start_offset + 2].try_into().unwrap();
+        u16::from_le_bytes(bytes)
+    }
+
+    #[inline]
+    #[allow(clippy::comparison_chain)]
+    // Looks for the element in the block. Returns the positions if found.
+    fn binary_search(&self, data: &[u8], target: u16) -> Option<u16> {
+        let mut size = self.num_vals;
+        let mut left = 0;
+        let mut right = size;
+        while left < right {
+            let mid = left + size / 2;
+
+            let mid_val = self.value_at_idx(data, mid);
+
+            if target > mid_val {
+                left = mid + 1;
+            } else if target < mid_val {
+                right = mid;
+            } else {
+                return Some(mid);
+            }
+
+            size = right - left;
+        }
+        None
+    }
+}
+
+const SERIALIZED_BLOCK_METADATA_SIZE: usize = 4;
+
+fn deserialize_sparse_codec_block(data: &[u8]) -> Vec<SparseCodecBlock> {
+    // The number of vals so far
+    let mut offset = 0;
+    let mut sparse_codec_blocks = Vec::new();
+    let num_blocks = u16::from_le_bytes([data[data.len() - 2], data[data.len() - 1]]);
+    let block_data_index_start =
+        data.len() - 2 - num_blocks as usize * SERIALIZED_BLOCK_METADATA_SIZE;
+    for block_num in 0..num_blocks as usize {
+        let block_data_index = block_data_index_start + SERIALIZED_BLOCK_METADATA_SIZE * block_num;
+        let block_idx = u16::from_le_bytes([data[block_data_index], data[block_data_index + 1]]);
+        let num_vals = u16::from_le_bytes([data[block_data_index + 2], data[block_data_index + 3]]);
+        while block_idx != sparse_codec_blocks.len() as u16 {
+            // fill up empty blocks
+            sparse_codec_blocks.push(SparseCodecBlock::empty_block(offset));
+        }
+        let block = SparseCodecBlock { num_vals, offset };
+        sparse_codec_blocks.push(block);
+        offset += num_vals as u32;
+    }
+    sparse_codec_blocks.push(SparseCodecBlock::empty_block(offset));
+    sparse_codec_blocks
+}
+
+const ELEMENTS_PER_BLOCK: u32 = u16::MAX as u32;
+
+/// Splits a idx into block index and value in the block
+fn split_in_block_idx_and_val_in_block(idx: u32) -> (u16, u16) {
+    let val_in_block = (idx % u16::MAX as u32) as u16;
+    let block_idx = (idx / u16::MAX as u32) as u16;
+    (block_idx, val_in_block)
```

Comment:
nitpick. I like to introduce a small `Copy` struct for those...

It makes the code more readable and avoids the positional result.
The consumes cannot intervert block_idx, or val_in_block.

Reply:
nitpick. I like to introduce a small `Copy` struct for those...

It makes the code more readable and avoids the positional result.
The consumes cannot intervert block_idx, or val_in_block.

```
struct ValueAddr {
    block_id: u16,
    value_in_block: u16,
}

fn value_addr(doc_id: u32) -> ValueAddr {
    const DIV: u32 = 1<<16;
    let val_in_block = (doc_id % DIV) as u16;
    let block_id = (doc_id / DIV) as u16;
    ValueAddr { block_id, val_in_block }
}
```